### ISSUE TITLE
Feature idea

### ISSUE THREAD

**Issue by justbur** [karthink/popper#6]
Thanks for the package. It works well so far. 

I was thinking that it might be nice if there was a setting that prevented the popup buffers from actually showing, but put an indicator in the mode line. For example, instead of a warning buffer popping up, this package puts an indicator in the mode line that there is a warning, and then I can bring up the buffer when I want.

Just a thought...

**Comment by karthink** [comment]
Interesting idea. How do you see it working exactly? If I have an open popup and a warning buffer is "suppressed", what happens when I 

- Call `popper-toggle-latest` twice?
- Call `popper-cycle` a few times?

**Comment by justbur** [comment]
What I was imagining is something like:

1. The popup buffer is not showing
2. A warning gets generated, which would normally show the popup buffer, but is instead suppressed. 
3. I don't see a new popup, but I do see an indicator in the mode line that a popup has been suppressed. 
4. I call `popper-toggle-latest` to see the warning buffer if I'm interested (and get rid of the indicator in the mode line about the unseen buffer), but otherwise I can go about doing what I wanted to. 

In other words, for some popups I would like to be notified that there is something new in them, but I'd like to be in control of when I see them. 

Maybe this is the wrong place to ask for this, but it seemed to be something that would be a natural fit with this package.

**Comment by karthink** [comment]
Thanks for the clarification. I understood the basic idea, I asked those specific questions because I'm trying to figure out how the popup toggling and cycling logic would need to change as a result.

1. What happens when cycling through popups? Are suppressed popups made visible?

2. Let's say there's a suppressed popup S and (non-suppressed) popup N appears on the frame. When I call `popper-toggle-latest` popup N disappears. When I call `popper-toggle-latest` again, does popup N appear or does S? If N appears, how do you access S? If S appears instead, then `popper-toggle-latest` is lying to the user in that it's not toggling the latest popup (which is N).

3. Because "popups" can appear as a result of any action in Emacs, the primary source for updating and maintaining popup information is counting visible popup windows on the frame. (This runs as part of window-configuration-change-hook.) A suppressed popup would never be counted and would need to be handled through a separate branch of the code. This might be worth doing if it can work seamlessly with visible popups.

I think the feature would be a neat fit too. I'm interested in implementing it but it'll help if you can lay out the intended usage and behavior in more detail. Feel free to submit a PR as well.

**Comment by justbur** [comment]
ok, great. 

Initially, I just had a very simple idea in mind (idea 1 below), but I can certainly imagine making it a little more complicated.

Idea 1. The only distinction made for S popups, as you call them, is that if the popup buffer is not showing, these S popups cannot force it open. In every other way, the S and N popups are treated as they are currently. S popups are visible when the user calls one of the functions to toggle or cycle, `popper-toggle-latest` would still bring up the S popup if it is the most recent, etc. 

Idea 2. Same as idea 1, except change the behavior when a popup is showing. If a popup is showing (regardless of the type) a suppressed popup is added to the "stack" but does not replace the currently visible popup. Maybe it just becomes the next one in the cycle? Otherwise, there are no other distinctions between popup types. 

Idea 3. Start to handle popup types differently behind the scenes. Maybe have a stack of S popups and a stack of N popups and have separate ways of cycling through them or even showing them. 

I think Idea 1 addresses the main annoyance I have, one of these popups which I don't immediately care about shows and messes up the window layout. Idea 2 adds to that idea a little. Idea 3 is more radical, and I'm not sure how cleanly it will work right now. 

I might work out a PR if I get some time. It's helpful to know that you're interested in it. I don't fully understand the way popups are handled in emacs so it would take some time to learn the details.

**Comment by karthink** [comment]
I'm on board with ideas 1 and 2. 3 is best handled by a different package.

> I don't fully understand the way popups are handled in emacs so it would take some time to learn the details.

Emacs has no concept of a "popup", this distinction is enforced entirely by Popper through (i) an internal stack of open and closed popups (`popper-open-popup-alist` and `popper-buried-popup-alist` respectively), and (ii) A custom rule in `display-buffer-alist` for how these special buffers should be displayed.

`popper-open-popup-alist` is only updated when the window configuration changes. Since a suppressed popup doesn't show up on the frame, it will "slip the net" and never be registered by Popper as a popup. I'll think about how to catch them in a predictable way.

**Comment by justbur** [comment]
ok, I tried this just now

```elisp
(setq popper-hidden-names '("\\*test\\*"))
(setq popper-hidden-modes nil)

(defun popper-hidden-buffer-p (buffer)
  (or (seq-some (lambda (buf-regexp) (string-match-p buf-regexp (buffer-name buffer)))
                popper-hidden-buffers)
      (member (buffer-local-value 'major-mode buf) popper-hidden-modes)))

(defun popper-select-popup-at-bottom-maybe-hide (buffer &optional _alist)
  (if (popper-hidden-buffer-p buffer)
      (display-buffer-no-window buffer nil)
    (let ((window (display-buffer-in-side-window
                   buffer
                   '((window-height . (lambda (win)
                                        (fit-window-to-buffer
                                         win
                                         (floor (frame-height) 3))))
                     (side . bottom)
                     (slot . 1)))))
      (select-window window))))

(setq popper-display-function #'popper-select-popup-at-bottom-maybe-hide)

(pop-to-buffer "*test*")
```

which seems like it should work based on the docstring of `display-buffer-no-window` but it doesn't. `display-buffer` seems to just fallback to `display-buffer-fallback-action` and display the buffer anyway. This actually appears to be a bug in `display-buffer-no-window` (it returns `nil` which indicates to `display-buffer` that another action should be tried). 

What does work is something like

```elisp
(setq popper-hidden-names '("\\*test\\*"))
(setq popper-hidden-modes nil)

(defun popper-hidden-buffer-p (buffer)
  (or (seq-some (lambda (buf-regexp) (string-match-p buf-regexp (buffer-name buffer)))
                popper-hidden-buffers)
      (member (buffer-local-value 'major-mode buf) popper-hidden-modes)))

(defun popper-display-buffer-function (buffer &optional alist)
  "docstring"
  (if (popper-hidden-buffer-p buffer)
      nil
    (let (display-buffer-function)
      (display-buffer buffer alist))))

(setq display-buffer-function 'popper-display-buffer-function)

(pop-to-buffer "*test*")
```

because `display-buffer` always accepts the return value of `display-buffer-function`.

**Comment by justbur** [comment]
Actually, digging a little deeper, this does work (with the alist entry in `display-buffer-no-window`)

```elisp
(add-to-list 'popper-reference-names "\\*test\\*")
(setq popper-hidden-names '("\\*test\\*"))
(setq popper-hidden-modes nil)

(defun popper-hidden-buffer-p (buffer)
  (or (seq-some (lambda (buf-regexp) (string-match-p buf-regexp (buffer-name buffer)))
                popper-hidden-names)
      (member (buffer-local-value 'major-mode buffer) popper-hidden-modes)))

(defun popper-select-popup-at-bottom-maybe-hide (buffer &optional _alist)
  (if (popper-hidden-buffer-p buffer)
      (display-buffer-no-window buffer '((allow-no-window . t)))
    (let ((window (display-buffer-in-side-window
                   buffer
                   '((window-height . (lambda (win)
                                        (fit-window-to-buffer
                                         win
                                         (floor (frame-height) 3))))
                     (side . bottom)
                     (slot . 1)))))
      (select-window window))))

(setq popper-display-function #'popper-select-popup-at-bottom-maybe-hide)

(pop-to-buffer "*test*")
```
